### PR TITLE
✨ Automatically generate CSP hashes for amp-script

### LINKED ISSUE #3608: amp-script demo on playground is near useless
### Current behavior
Due to the [playground](https://playground.amp.dev/?url=https%3A%2F%2Fpreview.amp.dev%2Fdocumentation%2Fexamples%2Fcomponents%2Famp-script) being a single file, the samples for amp-script are inlined. [Due to the hash requirements](https://amp.dev/documentation/components/amp-script/?format=websites#security-features), it means that any time a single byte is modified, the demo breaks with no obvious reason until you dig into the console errors.

### Expected behavior/code
You should be able to edit the amp-script demo without recompiling the hash every time

### PR DESCRIPTION
**Pull request by matthiasrohmer**
Fixes #3608. Currently, only for inline-scripts. Support for remote scripts (`<amp-script src="...">`) could be quickly added - just wasn't sure if it's safe to assume they are always fetchable...?

Hashes that are manually added are left untouched.

### CODE REVIEW

File: playground/src/csp-hash-calculator/csp-hash-calculator.js [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    const lineCount = this.editor.lineCount();
+
+    // Reset previous parsing results
+    this.head = {};
+    this.metaAmpScriptSrc = {};
+    this.scripts = [];
+
+    let i = 0;
+    let script = {from: null, to: null};
+    while (i < lineCount) {
```
Comment (sebastianbenz) [inline playground/src/csp-hash-calculator/csp-hash-calculator.js:112]: why not use a for loop here?

File: playground/src/csp-hash-calculator/csp-hash-calculator.js [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+      const tokens = this.editor.getLineTokens(i);
+
+      let j = 0;
+      while (j < tokens.length) {
```
Comment (sebastianbenz) [inline playground/src/csp-hash-calculator/csp-hash-calculator.js:116]: for loop?

File: playground/src/csp-hash-calculator/csp-hash-calculator.js [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+      this.metaAmpScriptSrc.to
+    );
+  }
+
+  _getExistingHashes() {
+    const html = this.editor.codeMirror.getRange(
+      this.metaAmpScriptSrc.from,
+      this.metaAmpScriptSrc.to
+    );
+    return html.match(/(?<=content=").*?(?=")/gms)[0].split(' ');
```
Comment (sebastianbenz) [inline playground/src/csp-hash-calculator/csp-hash-calculator.js:88]: This fails `npm run build:playground`:

```
ERROR in main.2eaca583ec3a65d6b302.js:51525 from closure-compiler: Cannot convert ECMASCRIPT_2018 feature "RegExp Lookbehind" to targeted output language.
      return html.match(/(?<=content=")[\s\S]*?(?=")/gm)[0].split(' ');
```

Can you also add this to the package.json (this got lost during the migration):

```
 "browserslist": [
   "last 2 versions", "not dead"
  ],
```

File: playground/src/csp-hash-calculator/csp-hash-calculator.js [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                .getRange(this.metaAmpScriptSrc.from, this.metaAmpScriptSrc.to)
+                .includes('name="amp-script-src"')
+            ) {
+              this.metaAmpScriptSrc = {};
+            }
+          }
+        }
+
+        // Look for script elements with target="amp-script" attribute
+        if (token.type == 'tag' && token.string == 'script') {
```
Comment (sebastianbenz) [inline playground/src/csp-hash-calculator/csp-hash-calculator.js:150]: extract into a method

File: playground/src/csp-hash-calculator/csp-hash-calculator.js [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+          if (!this.head.from) {
+            this.head.from = {line: i, ch: token.start - 1};
+          } else {
+            this.head.to = {line: i, ch: token.end + 1};
+          }
+        }
+
+        // Also check wether there is an existing meta[name="amp-script-src"]
+        // we could add to if none hasn't been found yet
+        if (!this.metaAmpScriptSrc.to) {
```
Comment (sebastianbenz) [inline playground/src/csp-hash-calculator/csp-hash-calculator.js:130]: extract into a method

File: playground/src/csp-hash-calculator/csp-hash-calculator.js [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+    let i = 0;
+    let script = {from: null, to: null};
+    while (i < lineCount) {
+      const tokens = this.editor.getLineTokens(i);
+
+      let j = 0;
+      while (j < tokens.length) {
+        const token = tokens[j];
+        // Save head position for to know where to add the meta element if
```
Comment (sebastianbenz) [inline playground/src/csp-hash-calculator/csp-hash-calculator.js:118]: extract into a method

File: playground/src/csp-hash-calculator/csp-hash-calculator.js (hunk 1 above)
Comment (matthiasrohmer) [reply playground/src/csp-hash-calculator/csp-hash-calculator.js]: Took what was in auto-importer.js and went with it 😁

**Comment by matthiasrohmer** [comment]
Thanks for the review! Accommodated all comments, PTAL 🙂

**Pull request merged**